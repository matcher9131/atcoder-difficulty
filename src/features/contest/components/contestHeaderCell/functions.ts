export const getTextColor = (maxRating: number | "inf"): string => {
    if (maxRating == "inf") {
        return "text-rating-2800";
    } else if (maxRating >= 2800) {
        return "bg-gradient-to-b from-rating-2800 from-50% to-rating-2400 to-50% text-transparent bg-clip-text";
    } else {
        switch (Math.floor(maxRating / 400)) {
            case 0:
                return "text-rating-0";
            case 1:
                return "text-rating-400";
            case 2:
                return "text-rating-800";
            case 3:
                return "text-rating-1200";
            case 4:
                return "text-rating-1600";
            case 5:
                return "text-rating-2000";
            case 6:
                return "text-rating-2400";
            default:
                return "text-rating-2800";
        }
    }
};
