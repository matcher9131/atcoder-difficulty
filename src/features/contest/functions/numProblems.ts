import type { ContestType } from "../types/contestType";

export const getNumProblems = (contestType: ContestType): number => {
    switch (contestType) {
        case "abc":
            return 8;
        case "arc":
            return 7;
        case "agc":
            return 10;
        default:
            throw new Error(`Unknown value: ${(contestType as { type: "__invalid__" }).type}`);
    }
};

export const getGridColsClassName = (contestType: ContestType): string => {
    switch (contestType) {
        case "abc":
            return "w-[112.5%] grid-cols-9";
        case "arc":
            return "w-[114%] grid-cols-8";
        case "agc":
            return "w-[157%] grid-cols-11";
        default:
            throw new Error(`Unknown value: ${(contestType as { type: "__invalid__" }).type}`);
    }
};
