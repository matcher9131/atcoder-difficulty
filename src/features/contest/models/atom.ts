import { atom } from "jotai";
import contestUrl from "../../../assets/contests.json?url";
import { parseDateOrNull } from "../../../utils/date";
import type { Contest } from "../types/contest";

type ContestsJson = {
    readonly body: Record<string, { readonly d: string; readonly m: number | "inf" }>;
    readonly lastOfChunks: readonly string[];
};

const loadContests = async (): Promise<ContestsJson> => {
    const response = await fetch(contestUrl);
    if (!response.ok) throw new Error("Failed loading contests.");
    return (await response.json()) as ContestsJson;
};

export const contestsJsonAtom = atom(await loadContests());

export const contestsAtom = atom((get) =>
    Object.entries(get(contestsJsonAtom).body)
        .flatMap(([id, { d, m }]) => {
            const date = parseDateOrNull(d);
            if (date == null) return [];
            return [{ id, d: date, m }];
        })
        .toSorted((x: Contest, y: Contest) => y.d.getTime() - x.d.getTime()),
);
