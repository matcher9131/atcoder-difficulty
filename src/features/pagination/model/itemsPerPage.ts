import { atom } from "jotai";
import type { PaginationKey } from "../types/paginationKey";

const abcItemsPerPageAtom = atom(20);
const arcItemsPerPageAtom = atom(20);
const agcItemsPerPageAtom = atom(20);
const solveProbabilityItemsPerPageAtom = atom(20);

export const itemsPerPageAtom = (key: PaginationKey): ReturnType<typeof atom<number>> => {
    switch (key) {
        case "abc":
            return abcItemsPerPageAtom;
        case "arc":
            return arcItemsPerPageAtom;
        case "agc":
            return agcItemsPerPageAtom;
        case "solveProbability":
            return solveProbabilityItemsPerPageAtom;
        default:
            throw new Error(`Unknown key: ${(key as { type: "__invalid__" }).type}`);
    }
};
