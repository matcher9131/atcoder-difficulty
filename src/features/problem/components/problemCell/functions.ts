export const getFillColor = (rating: number): string => {
    switch (Math.floor(rating / 400)) {
        case 0:
            return "fill-rating-0";
        case 1:
            return "fill-rating-400";
        case 2:
            return "fill-rating-800";
        case 3:
            return "fill-rating-1200";
        case 4:
            return "fill-rating-1600";
        case 5:
            return "fill-rating-2000";
        case 6:
            return "fill-rating-2400";
        default:
            return "fill-rating-2800";
    }
};

export const getTextColor = (rating: number): string => {
    switch (Math.floor(rating / 400)) {
        case 0:
            return "text-rating-0";
        case 1:
            return "text-rating-400";
        case 2:
            return "text-rating-800";
        case 3:
            return "text-rating-1200";
        case 4:
            return "text-rating-1600";
        case 5:
            return "text-rating-2000";
        case 6:
            return "text-rating-2400";
        default:
            return "text-rating-2800";
    }
};
