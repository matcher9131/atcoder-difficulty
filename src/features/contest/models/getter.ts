import { atom } from "jotai";
import { atomFamily } from "jotai/utils";
import type { ContestType } from "../types/contestType";
import { contestsAtom, contestsJsonAtom } from "./atom";

const contestAtom = atomFamily((contestId: string) =>
    atom((get) => get(contestsAtom).find((contest) => contest.id === contestId)),
);

export const contestDateAtom = atomFamily((contestId: string) =>
    atom((get) => {
        const contest = get(contestAtom(contestId));
        if (contest == null) throw new Error(`Invalid contestId: ${contestId}`);
        return contest.d;
    }),
);

export const contestMaxRatingAtom = atomFamily((contestId: string) =>
    atom((get) => {
        const contest = get(contestAtom(contestId));
        if (contest == null) throw new Error(`Invalid contestId: ${contestId}`);
        return contest.m;
    }),
);

const abcLikeContestIdsAtom = atom((get) =>
    get(contestsAtom).flatMap((contest) => (contest.m !== "inf" && contest.m < 2000 ? [contest.id] : [])),
);

const arcLikeContestIdsAtom = atom((get) =>
    get(contestsAtom).flatMap((contest) => (contest.m !== "inf" && contest.m >= 2000 ? [contest.id] : [])),
);

const agcLikeContestIdsAtom = atom((get) =>
    get(contestsAtom).flatMap((contest) => (contest.m === "inf" ? [contest.id] : [])),
);

export const contestIdsByTypeAtom = (contestType: ContestType) => {
    switch (contestType) {
        case "abc":
            return abcLikeContestIdsAtom;
        case "arc":
            return arcLikeContestIdsAtom;
        case "agc":
            return agcLikeContestIdsAtom;
        default:
            throw new Error(`Unknown value: ${(contestType as { type: "__invalid__" }).type}`);
    }
};

export const contestStatsDateLastOfChunkAtom = atom((get) =>
    get(contestsJsonAtom).lastOfChunks.map((contestId) => get(contestDateAtom(contestId))),
);
