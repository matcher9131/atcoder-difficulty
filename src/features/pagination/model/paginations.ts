import { atom, type Atom } from "jotai";
import { contestIdsByTypeAtom } from "../../contest/dict/contests";
import type { PaginationKey } from "../types/paginationKey";
import { solveProbabilitiesMiddleIndexAtom } from "../../solveProbability/models/solveProbabilities";
import { numProblemsAtom } from "../../problem/dict/problems";

const abcPaginationValueAtom = atom(0);
const arcPaginationValueAtom = atom(0);
const agcPaginationValueAtom = atom(0);

export const paginationValueAtom = (key: PaginationKey): ReturnType<typeof atom<number>> => {
    switch (key) {
        case "abc":
            return abcPaginationValueAtom;
        case "arc":
            return arcPaginationValueAtom;
        case "agc":
            return agcPaginationValueAtom;
        default:
            throw new Error(`Unknown key: ${(key as { type: "__invalid__" }).type}`);
    }
};

const abcPaginationMaxValueAtom = atom((get) => Math.ceil(get(contestIdsByTypeAtom("abc")).length / 100));
const arcPaginationMaxValueAtom = atom((get) => Math.ceil(get(contestIdsByTypeAtom("arc")).length / 100));
const agcPaginationMaxValueAtom = atom((get) => Math.ceil(get(contestIdsByTypeAtom("agc")).length / 100));

export const paginationMaxValueAtom = (key: PaginationKey): Atom<number> => {
    switch (key) {
        case "abc":
            return abcPaginationMaxValueAtom;
        case "arc":
            return arcPaginationMaxValueAtom;
        case "agc":
            return agcPaginationMaxValueAtom;
        default:
            throw new Error(`Unknown key: ${(key as { type: "__invalid__" }).type}`);
    }
};

export const solveProbabilityPaginationValueAtom = atom(0);
export const solveProbabilityPaginationMinValueAtom = atom((get) => {
    const l = get(solveProbabilitiesMiddleIndexAtom) - 50;
    return -Math.ceil(l / 100);
});
export const solveProbabilityPaginationMaxValueAtom = atom((get) => {
    const r = get(solveProbabilitiesMiddleIndexAtom) + 50;
    return Math.ceil((get(numProblemsAtom) - r) / 100) + 1;
});
