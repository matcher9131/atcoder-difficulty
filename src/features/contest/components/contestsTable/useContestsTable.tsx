import { useAtom } from "jotai";
import type { ContestType } from "../../types/contestType";
import { contestIdsByTypeAtom } from "../../dict/contests";
import type { ContestsTableProps } from "./ContestsTable";
import { ContestRowContainer } from "../contestRow";
import { getGridColsClassName, getNumProblems } from "../../functions/numProblems";
import { paginationValueAtom } from "../../../pagination/model/paginations";
import { ContestsTableHeader } from "../contestsTableHeader";
import { itemsPerPageAtom } from "../../../pagination/model/itemsPerPage";

export const useContestsTable = (contestType: ContestType): ContestsTableProps => {
    const numProblems = getNumProblems(contestType);
    const contestsTableHeader = <ContestsTableHeader numProblems={numProblems} />;
    const gridColsClassName = getGridColsClassName(contestType);
    const [contestIds] = useAtom(contestIdsByTypeAtom(contestType));
    const [pageIndex] = useAtom(paginationValueAtom(contestType));
    const [itemsPerPage] = useAtom(itemsPerPageAtom(contestType));
    const contestRows = contestIds
        .slice(itemsPerPage * pageIndex, itemsPerPage * (pageIndex + 1))
        .map((contestId) => <ContestRowContainer key={contestId} contestId={contestId} numProblems={numProblems} />);
    return { contestsTableHeader, gridColsClassName, contestRows };
};
