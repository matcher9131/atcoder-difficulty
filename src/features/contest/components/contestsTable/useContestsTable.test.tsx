import { describe, it, expect, vi, beforeEach } from "vitest";
import { render, renderHook } from "@testing-library/react";
import { useContestsTable } from "./useContestsTable";
import { useAtom } from "jotai";
import React from "react";
import { getNumProblems, getGridColsClassName } from "../../functions/numProblems";
import { contestIdsByTypeAtom } from "../../models/getter";
import { paginationValueAtom } from "../../../pagination/model/paginations";
import { itemsPerPageAtom } from "../../../pagination/model/itemsPerPage";

vi.mock("jotai", () => ({
    useAtom: vi.fn(),
}));
vi.mock("../../dict/contests", () => ({
    contestIdsByTypeAtom: vi.fn(),
}));
vi.mock("../../../pagination/model/paginations", () => ({
    paginationValueAtom: vi.fn(),
}));
vi.mock("../../../pagination/model/itemsPerPage", () => ({
    itemsPerPageAtom: vi.fn(),
}));
vi.mock("../../functions/numProblems", () => ({
    getNumProblems: vi.fn(),
    getGridColsClassName: vi.fn(),
}));
vi.mock("../contestRow", () => ({
    ContestRowContainer: vi.fn(
        ({ contestId, numProblems }: { readonly contestId: string; readonly numProblems: number }) => (
            <tr data-testid="contest-row" data-contestid={contestId} data-numproblems={numProblems} />
        ),
    ),
}));
vi.mock("../contestsTableHeader", () => ({
    ContestsTableHeader: vi.fn(({ numProblems }: { readonly numProblems: number }) => (
        <thead data-testid="table-header" data-numproblems={numProblems} />
    )),
}));

describe("useContestsTable", () => {
    const dummyContestType = "abc" as const;
    const dummyNumProblems = 4;
    const dummyGridColsClassName = "grid-cols-4";
    const dummyContestIds = ["c1", "c2", "c3", "c4", "c5"];
    const dummyPageIndex = 1;
    const dummyItemsPerPage = 2;

    beforeEach(() => {
        vi.clearAllMocks();
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
        (getNumProblems as any).mockReturnValue(dummyNumProblems);
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
        (getGridColsClassName as any).mockReturnValue(dummyGridColsClassName);

        // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
        (contestIdsByTypeAtom as any).mockImplementation(() => [dummyContestIds]);
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
        (paginationValueAtom as any).mockImplementation(() => [dummyPageIndex]);
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
        (itemsPerPageAtom as any).mockImplementation(() => [dummyItemsPerPage]);
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
        (useAtom as any).mockImplementation((value: unknown) => value);
    });

    it("returns correct props structure", () => {
        const { result } = renderHook(() => useContestsTable(dummyContestType));
        expect(result.current).toHaveProperty("contestsTableHeader");
        expect(result.current).toHaveProperty("gridColsClassName", dummyGridColsClassName);
        expect(result.current).toHaveProperty("contestRows");
        expect(Array.isArray(result.current.contestRows)).toBe(true);
    });

    it("renders ContestsTableHeader with correct numProblems", () => {
        const { result } = renderHook(() => useContestsTable(dummyContestType));
        const { contestsTableHeader } = result.current;
        const { getByTestId } = render(contestsTableHeader);
        expect(getByTestId("table-header")).toHaveAttribute("data-numproblems", String(dummyNumProblems));
    });

    it("renders ContestRowContainer for paginated contestIds", () => {
        // pageIndex=1, itemsPerPage=2 => slice(2,4) => ["c3","c4"]
        const { result } = renderHook(() => useContestsTable(dummyContestType));
        const { contestRows } = result.current;
        const { getAllByTestId } = render(<>{contestRows.map((cell, i) => React.cloneElement(cell, { key: i }))}</>);
        const rows = getAllByTestId("contest-row");
        expect(rows).toHaveLength(2);
        expect(rows[0].getAttribute("data-contestid")).toBe("c3");
        expect(rows[1].getAttribute("data-contestid")).toBe("c4");
    });

    it("returns empty contestRows if contestIds is empty", () => {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
        (contestIdsByTypeAtom as any).mockImplementation(() => [[]]);
        const { result } = renderHook(() => useContestsTable(dummyContestType));
        expect(result.current.contestRows).toHaveLength(0);
    });

    it("matches snapshot", () => {
        const { result } = renderHook(() => useContestsTable(dummyContestType));
        expect(result).toMatchSnapshot();
    });
});
