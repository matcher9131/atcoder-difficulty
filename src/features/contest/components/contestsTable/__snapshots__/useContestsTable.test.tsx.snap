// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useContestsTable > matches snapshot 1`] = `
{
  "current": {
    "contestRows": [
      <spy
        contestId="c3"
        numProblems={4}
      />,
      <spy
        contestId="c4"
        numProblems={4}
      />,
    ],
    "contestsTableHeader": <spy
      numProblems={4}
    />,
    "gridColsClassName": "grid-cols-4",
  },
}
`;
